### PR TITLE
Parse Tree Implementation

### PR DESCRIPTION
**Pull request by schuylermartin45**
This is the initial parse tree work. There is a lot here, so here are some highlights:

**It has:**
- A recipe parser that can track comments/selectors, Jinja variables, and values through look-up tables and a parse tree.
- Support for `replace` and `test` that closely (if not completely) follows the JSON Patch RPC requirements
- 18 Test cases. I believe there is at least 1 test case for every publicly available function that is completed. Tests pass locally with `make tests`
- Ability to:
  - Render recipes
  - Ability to traverse the tree structure with an underlying set of private traversal functions
  - Get/Set/Find variables and values in the data structure
  - Get selectors in the data structure
  - Validate incoming JSON Patch requests. Fields that are contingent on the `op` type are validated.

**It does not have:**
- Wiring into the existing `percy` framework. In other words, merging this will not break anything existing. This class is separate from the current `recipe.py` file.
- Code that is validated with `mypy` and `pylint`
- Support for the other 4 JSON Patch verbs
- Fixes for a handful of known bugs, tracked in the JIRA epic (see below)
- Ability to generically search for a regex statement across the parse tree and all variables/selectors (this is in a future ticket)

Known bugs and feature improvements are being tracked in our board as JIRA Epic [PKG-2787](https://anaconda.atlassian.net/browse/PKG-2787)

[PKG-2787]: https://anaconda.atlassian.net/browse/PKG-2787?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by schuylermartin45** [comment]
> Now some of this work uses 3.11/newer Python syntax. I don't think that should be an issue with the older versions as again, this isn't linked into any existing code and is a stand-alone file. We also talked about bumping our min version last week that would solve this issue.

Well, okay, if the tests are set to run in GitHub actions, I guess that'd execute code that may not be compatible.

**Comment by schuylermartin45** [comment]
I have been experimenting with incorporating the `mypy` and `pylint` learnings from `perseverance-scripts` into GitHub Actions today [on this branch](https://github.com/anaconda-distribution/percy/tree/schuy_recipe_parser_with_github_actions_changes).

Once I finish the last few issues, I plan to merge those configuration changes into this PR. Or I'll make a second PR when this one gets merged.

File: percy/tests/test_recipe_parser.py [hunk 1]
```diff
@@ -0,0 +1,475 @@
...
+    assert parser.get_value("/build/number") == { "number": 0 }
+    assert parser.get_value("/build/number/") == 0
+    # Return a compound value
+    assert parser.get_value("/build") == { "number": 0, "skip": True, "is_true": True }
+    assert parser.get_value("/build/") == { "number": 0, "skip": True, "is_true": True }
+    # Return a value in a list
+    assert parser.get_value("/requirements/host") == ["setuptools", "fakereq"]
+    assert parser.get_value("/requirements/host/") == ["setuptools", "fakereq"]
+    assert parser.get_value("/requirements/host/0") == "setuptools"
+    assert parser.get_value("/requirements/host/1") == "fakereq"
```
Comment (cbouss) [inline percy/tests/test_recipe_parser.py:108]: We don't always know the index of a dependency.
I would expect this:
```
host:
  - numpy 1.21 # [py<311]
  - numpy 1.23 # [py==311]
```
to allow this kind of queries:
`/requirements/host/numpy` --> `[("numpy", "1.21", "py<311"), ("numpy", "1.23", "py<==311")]`

And a corresponding set_value to allow the opposite.

An example use case is for supporting a new python version. 
In that case I would want to transform the initial host part to:
```
host:
  - numpy 1.21 # [py<311]
  - numpy 1.23 # [py==311]
  - numpy 1.26 # [py==312]
```
Or even better, if know the inital part matches the cbc, to:
```
host:
  - numpy {{ numpy }}
```

Comment (schuylermartin45) [reply percy/tests/test_recipe_parser.py:108]: This is a really good call-out. 

To query `.../numpy` would break the RFC specification for JSON-patch. I'm generally not a fan of being super reliant on string data as input to a function (like how some folks don't like having raw SQL queries in code), but I do find that the JSON-patch system is very useful for developers.

I guess what I'm saying is I'd prefer to solve this case another way, with specific function calls that handle the processing of dependencies. I think that'll closely tie into the [pinning work](https://anaconda.atlassian.net/browse/PKG-2833) I have planned, so I'm going to comment in the JIRA ticket to handle this case selector case.

I haven't thought it out super well, but I imagine that it'd be super useful to have a look-up table that tracks dependencies and what the pinnings/selectors are set to. That table could then reference nodes individually. I have something similar for selectors already, with functions that allow the user to query what selectors are available.

I also have plans for a general regex search tool that'll match a pattern against the entire tree. The result would be a set of paths to all the nodes that match.

Does that sound good to you @cbouss ?

Comment (schuylermartin45) [reply percy/tests/test_recipe_parser.py:108]: I figured that regex search tool would be able to cover a lot of edge cases that wouldn't be easily handled by the other member functions.

File: percy/render/recipe_parser.py [hunk 2]
```diff
@@ -0,0 +1,1195 @@
...
+        Remove a variable from the project. If one is not found, no changes
+        are made.
+        :param var: Variable to delete
+        """
+        if not var in self._vars_tbl:
+            return
+        del self._vars_tbl[var]
+        self._is_modified = True
+
+    def get_var_paths(self, var: str) -> list[str]:
```
Comment (JeanChristopheMorinPerso) [inline percy/render/recipe_parser.py:955]: Maybe `get_var_usages` or `get_var_references`?

File: percy/render/recipe_parser.py [hunk 3]
```diff
@@ -0,0 +1,1195 @@
...
+        # NOTE: Traversing the tree and generating our own data structures will
+        # be more efficient than rendering and leveraging the YAML parser, BUT
+        # this method re-uses code and is easier to maintain.
+        lst = []
+        RecipeParser._render_tree(node, -1, lst)
+        return RecipeParser._parse_yaml("\n".join(lst))
+
+    ## Jinja Variable Functions ##
+
+    def list_vars(self) -> list[str]:
```
Comment (JeanChristopheMorinPerso) [inline percy/render/recipe_parser.py:902]: Minor: `list_vars` vs `list_selectors`. One uses an abbreviation while the other doesn't. It makes the API inconsistent.

File: percy/render/recipe_parser.py [hunk 4]
```diff
@@ -0,0 +1,1195 @@
...
+        """
+        # Validate the patch schema
+        try:
+            schema_validate(patch, JSON_PATCH_SCHEMA)
+        except Exception as e:
+            raise JsonPatchValidationException(patch) from e
+
+        path: Final[str] = patch["path"]
+        # Path must have length
+        if not path:
```
Comment (JeanChristopheMorinPerso) [inline percy/render/recipe_parser.py:1139]: This can be enforced directly with the JSON Schema by using [minLength](https://json-schema.org/draft/2020-12/json-schema-validation.html#name-minlength).

**Review by JeanChristopheMorinPerso — COMMENTED** [review]
I left a first pass of comments. Some of them are more nitpicks than others. Some relates to how the API looks like (which I think are more important than nitpicks). I did not spend a lot of time to think about the functionality it currently has or how things are implemented though.

File: percy/render/recipe_parser.py (hunk 4 above)
Comment (schuylermartin45) [reply percy/render/recipe_parser.py:1139]: TIL. Nice!

File: percy/render/recipe_parser.py [hunk 5]
```diff
@@ -0,0 +1,1195 @@
...
+        return RecipeParser._parse_yaml("\n".join(lst))
+
+    ## Jinja Variable Functions ##
+
+    def list_vars(self) -> list[str]:
+        """
+        Returns variables found in the recipe, sorted by first appearance.
+        :return: List of variables found in the recipe.
+        """
+        return list(dict.fromkeys(self._vars_tbl))
```
[the remark this answers is not in the document]
Comment (schuylermartin45) [reply percy/render/recipe_parser.py:907]: From the docs:
```
The objects returned by [dict.keys()](https://docs.python.org/3/library/stdtypes.html#dict.keys), [dict.values()](https://docs.python.org/3/library/stdtypes.html#dict.values) and [dict.items()](https://docs.python.org/3/library/stdtypes.html#dict.items) are view objects. They provide a dynamic view on the dictionary’s entries, which means that when the dictionary changes, the view reflects these changes.
```
I _think_ the view objects are more or less analogous to generators so I'd still need to convert to a `list`

This was a recommended way of preserving key order (based on when the keys were added). I have a feeling the post is a little old as Python's dictionaries have maintained order since 3.6. I'm guessing I can simplify to 
```python
return list(self._vars_tbl.keys())
```

Also I'm not entirely sure if sort-order-by-first-appearance is super important, but it felt like a nice guarantee to have. I primarily enforced it for two reasons:
1) Ease of unit testing against lists
2) I figured it was a good guarantee to have for file traversals.

File: percy/render/recipe_parser.py [hunk 6]
```diff
@@ -0,0 +1,1195 @@
...
+        at that path.
+
+        If no Node is found at that path, return `None`.
+        :param node:    Starting node of the tree/branch to traverse.
+        :param path:    Path, as a stack, that describes a location in the tree.
+        :return: `_Node` object if a node is found in the parse tree at that
+                 path. Otherwise returns `None`.
+        """
+        # Bootstrap recursive edge cases
+        if _Node is None:
```
[the remark this answers is not in the document]
Comment (schuylermartin45) [reply percy/render/recipe_parser.py:296]: Good catch, that's a typo.

File: percy/render/recipe_parser.py [hunk 7]
```diff
@@ -0,0 +1,1195 @@
...
+
+    def __eq__(self, other: object) -> bool:
+        """
+        Checks if two recipe representations match entirely
+        :param other:   Other recipe parser instance to check against.
+        :return: True if both recipes contain the same current state. False
+                 otherwise.
+        """
+        if not isinstance(other, RecipeParser):
+            return NotImplemented
```
[the remark this answers is not in the document]
Comment (schuylermartin45) [reply percy/render/recipe_parser.py:721]: Should be `raise` not `return`. I would think that if the functionality isn't implemented yet, it'd be best to keep that exception over calling it a `TypeError`.

This is all only temporary so I guess it doesn't really matter. Once all `ops` are added, this check will be removed.

File: percy/render/recipe_parser.py [hunk 8]
```diff
@@ -0,0 +1,1195 @@
...
+                selector = match.group(0)
+                selector_info = SelectorInfo(node, path)
+                if selector not in self._selector_tbl:
+                    self._selector_tbl[selector] = [selector_info]
+                else:
+                    self._selector_tbl[selector].append(selector_info)
+
+        _Traverse.traverse_all(self._root, _collect_selectors)
+
+    def __init__(self, content: str):
```
[the remark this answers is not in the document]
Comment (schuylermartin45) [reply percy/render/recipe_parser.py:604]: This class does `render()` and allow queries against the parsed data. I will say I think the class file is a little big right now. I'm not immediately opposed to splitting up the file a bit, but I initially kept it all in one so the static analyzer wouldn't get angry for accessing private classes outside `recipe_parser.py`

File: percy/render/recipe_parser.py (hunk 3 above)
Comment (schuylermartin45) [reply percy/render/recipe_parser.py:902]: I don't think there's a common short-hand for `selectors`, but I see and agree where you're coming from.

File: percy/render/recipe_parser.py [hunk 9]
```diff
@@ -0,0 +1,1195 @@
...
+    def contains_var(self, var: str) -> bool:
+        """
+        Determines if a variable is set in this recipe.
+        :param var: Variable to check for.
+        :return: True if a variable name is found in this recipe. False
+                 otherwise.
+        """
+        return var in self._vars_tbl
+
+    def get_var(self, var: str, default: Primitives = None) -> Primitives:
```
[the remark this answers is not in the document]
Comment (schuylermartin45) [reply percy/render/recipe_parser.py:918]: I think I'm going to need to circle back with you on this one, I'm not sure if I follow.

File: percy/render/recipe_parser.py [hunk 10]
```diff
@@ -0,0 +1,1195 @@
...
+                return node
+            return None
+        # Purge `root` from the path
+        path.pop()
+        return _Traverse._traverse_recurse(node, path)
+
+    @staticmethod
+    def traverse_all(
+        node: _Node | None,
+        func: callable,
```
[the remark this answers is not in the document]
Comment (schuylermartin45) [reply percy/render/recipe_parser.py:311]: Should of put a TODO in for these. I got lazy. I think I'm also going to write a Type Alias as this callback signature gets used a lot.

File: percy/render/recipe_parser.py [hunk 11]
```diff
@@ -0,0 +1,1195 @@
...
+                _Node(
+                    value=value.splitlines(),
+                    is_multiline=True,
+                )
+            ]
+
+        # For complex types, generate the YAML equivalent and build
+        # a new tree.
+        if isinstance(value, (dict, list)):
+            return RecipeParser(  # pylint: disable=W0212
```
[the remark this answers is not in the document]
Comment (schuylermartin45) [reply percy/render/recipe_parser.py:570]: I've found that the IDs are significantly easier to google/look-up against the pylint list of errors, but that's just a personal preference.

File: percy/render/recipe_parser.py [hunk 12]
```diff
@@ -0,0 +1,1195 @@
...
+
+        # Read the YAML line-by-line, maintaining a stack to manage the last
+        # owning node in the tree.
+        node_stack: list[_Node] = [self._root]
+        # Relative depth is determined by the increase/decrease of indentation
+        # marks (spaces)
+        cur_indent = 0
+        last_node = node_stack[-1]
+        # Regex for detecting the start of a multiline node
+        multiline_re = re.compile(r"^\s*.*:\s+\|(\s*|\s+#.*)")
```
[the remark this answers is not in the document]
Comment (schuylermartin45) [reply percy/render/recipe_parser.py:638]: Good to know. I think `|` is all I've ever seen in our recipes. `>` would not be that difficult to implement, just look for that and `|`.

I think I can quickly add `>` support before this gets merged.

File: percy/render/recipe_parser.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (JeanChristopheMorinPerso) [reply percy/render/recipe_parser.py:570]: We write the code once, but it'll be read much more than that. Hence why using names is useful.

File: percy/render/recipe_parser.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (JeanChristopheMorinPerso) [reply percy/render/recipe_parser.py:907]: Dict order based on insertion is guaranteed since Python 3.7. It was initially added in 3.6 but it was an implementation detail and the order changed between 3.6 and 3.7. See https://docs.python.org/3.11/library/stdtypes.html#typesmapping:~:text=Changed%20in%20version%203.7%3A%20Dictionary%20order%20is%20guaranteed%20to%20be%20insertion%20order.%20This%20behavior%20was%20an%20implementation%20detail%20of%20CPython%20from%203.6.

File: percy/render/recipe_parser.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (JeanChristopheMorinPerso) [reply percy/render/recipe_parser.py:918]: I'm not even 100% sure that I really meant 😛 I think we can split this in 2:

* >The default's default should be a sentinel or something like that so that we can do `get_var("myvar", default=None)`.

  This would allow to have a behavior more similar (but not identical) to `dict.get`. THIs would be shorter than having to call `list_vars` to see if a variable exist and then get the value with `get_var`.
* >I have a preference for the non abbreviated form.

  Here I think I meant `get_variables` over `get_vars`.

File: percy/render/recipe_parser.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (schuylermartin45) [reply percy/render/recipe_parser.py:728]: Didn't realize this was in `__eq__`. Yeah TypeError makes sense

File: percy/render/recipe_parser.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (schuylermartin45) [reply percy/render/recipe_parser.py:925]: I'm still confused on that first point. I'm guessing `sentinel` has some python-specific meaning I haven't come across before.

I based this off of `dict.get()`, I'm not sure what the difference is. Can't you call an optional parameter by name in python?

File: percy/render/recipe_parser.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (schuylermartin45) [reply percy/render/recipe_parser.py:914]: Ah, wait a minute. A previous comment sheds light on why I did this. The comment just isn't on every example:
```
        # The list should be de-duped and maintain order. Duplications occur
        # when key-value pairings mean a selector occurs on two nodes with
        # the same path.
        #
        # For example:
        #   skip: True  # [unix]
        # The nodes for both `skip` and `True` contain the comment `[unix]`
```
Duplication occurs in some of these `fromkeys()` usages, but not all. I will act accordingly.
The trials and tribulations of blindly copy-pasting previous work.

File: percy/render/recipe_parser.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (schuylermartin45) [reply percy/render/recipe_parser.py:645]: ...I spoke too soon without reading the Stackoverflow post. I didn't think there was nuance here.

I'm going to cut a JIRA ticket to track this request.

[the remark this answers is not in the document]
Comment (schuylermartin45) [reply percy/render/recipe_parser.py:645]: https://anaconda.atlassian.net/browse/PKG-2868

File: percy/render/recipe_parser.py [hunk 13]
```diff
@@ -0,0 +1,1195 @@
...
+        :param comment:         Comment on the line this node was found on
+        :param children:        List of children nodes, descendants of this node
+        :param is_list_member:  Indicates if this node is part of a list
+        :param is_multiline:    Indicates if the node represents a multiline
+                                value.
+        """
+        self.value = value
+        self.comment = comment
+        self.children: list[_Node] = children if children else []
+        self.is_list_member = is_list_member
```
[the remark this answers is not in the document]
Comment (schuylermartin45) [reply percy/render/recipe_parser.py:197]: Better yet, I'll go with `list_member_flag` and `multiline_flag`. I think that gets the point across while being consistent.

File: percy/render/recipe_parser.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (JeanChristopheMorinPerso) [reply percy/render/recipe_parser.py:649]: YAML is a very nuanced and complex spec! I refrained from commenting on some other parts of the code that wouldn't handle other YAML cases because of that reason. But it's something to keep in mind.

**Comment by schuylermartin45** [comment]
I ran `pytest` locally against `3.9` to validate the type annotations in [7742a4f](https://github.com/anaconda-distribution/percy/pull/47/commits/7742a4f6b8165faac0626d81da0debad2573848a)

File: percy/render/recipe_parser.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (JeanChristopheMorinPerso) [reply percy/render/recipe_parser.py]: See https://peps.python.org/pep-0661/ for an explanation of what a sentinel is. This PEP isn't introducing the concept (since that concept has existed since the very early days of python) and I'm pointing to this PEP simply because it has a good explanation.

A common sentinel value in Python looks like this:
```python
_sentinel = object()


def myfunc(arg1, optional1=_sentinel):
    if optional1 == _sentinel:
        # The user didn't pass anything to the `optional1` kwarg
    else:
        # The user has passed someting
```
This allows a user to pass `None` and the function is still able to differenciate between "no value was provided" and "None was explicitly provided".

File: percy/render/recipe_parser.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (schuylermartin45) [reply percy/render/recipe_parser.py:649]: And that's why I backed this all with unit tests :) I'm very confident in how much I don't know in the YAML spec.

File: percy/render/recipe_parser.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (schuylermartin45) [reply percy/render/recipe_parser.py]: Ahhhhhhhhh I see it now, that's a very very subtle nuance in functionality.

File: percy/render/recipe_parser.py [hunk 14]
```diff
@@ -0,0 +1,1208 @@
...
+        self._is_modified = False
+
+        # Tracks Jinja variables set by the file
+        self._vars_tbl: dict[str, JsonType] = {}
+        # Find all the set statements and record the values
+        set_line_re = re.compile(r"{%\s*set.*=.*%}\s*\n")
+        for line in set_line_re.findall(self._init_content):
+            key = line[line.find("set") + len("set") : line.find("=")].strip()
+            value = line[line.find("=") + len("=") : line.find("%}")].strip()
+            self._vars_tbl[key] = ast.literal_eval(value)
```
Comment (JeanChristopheMorinPerso) [inline percy/render/recipe_parser.py:636]:
```suggestion
            try:
                self._vars_tbl[key] = ast.literal_eval(value)
            except:
                self._vars_tbl[key] = value
```
Would support any value and would still give you a python type other than string for simple values.

File: percy/render/recipe_parser.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (schuylermartin45) [reply percy/render/recipe_parser.py]: Replaced everywhere

**Review by JeanChristopheMorinPerso — APPROVED** [review]
Great work @schuylermartin45. It'll be interesting to see where this goes!

**Comment by schuylermartin45** [comment]
Talked to Mark and Charles offline, we are good to merge!
